middle_end/flambda2/types/env/typing_env.rec.ml

```diff
@@ -1050,6 +1052,22 @@ and add_equation t name ty =
          any name modes). *)
       let canonical = Aliases.get_canonical_ignoring_name_mode aliases name in
       canonical, t, ty
+    | alias_of when Simple.is_foreign_or_constant alias_of
+                 && Name.is_foreign name ->
+      (* At the point of writing we do not properly merge alias structures in
+         packs, and rely on the domains of their [canonical_elements] maps
+         being disjoint. In practice, this means that this map must only have
+         elements from the current compilation unit as keys.
+         It's already impossible for constants to end up as keys, as they're
+         always canonical. The same also holds for symbols.
+         A foreign variable can't alias to a non-foreign symbol either, but
+         the implementation doesn't know it; fortunately until we start
+         supporting arbitrary aliases (for example, as created by the physical
+         equality primitive), it's impossible to reach this case in practice.
```

Comment:
Could you please add a comment in the phys-equal simplification code as to the things we need to watch out for when implementing that properly?  I remember there was a patch fairly recently that would also have required modification if such an implementation existed, but I can't recall exactly what it was.